### PR TITLE
Add encoding preserving serialization for vectors

### PR DESCRIPTION
**Pull request by mbasmanova**
Add VectorSaver utility to save vectors with possibly multiple layers of
encodings to disk and restore these back in a different process.

This PR includes support for flat, constant and dictionary vectors of 
scalar and complex data types.

We can integrate this tool with expression evaluation to save the input vector 
on exception. We already use ExceptionContextSetter to add expression text 
in the Context field of the exception. We can extend it to also include the path 
of the file that has the input data. Here is an example:

```
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: (8 vs. 12) Malformed dictionary, index array is shorter than DictionaryVector
Retriable: False
Expression: dictionaryIndices->size() >= length * sizeof(vector_size_t)
Context: concat(cast((c0) as VARCHAR), ,:VARCHAR). Input: /tmp/velox_vector_f7dneH.
Function: DictionaryVector
File: .../velox/vector/DictionaryVector-inl.h
Line: 107
```

“Input: /tmp/velox_vector_f7dneH” in the Context: field shows the file path that 
contains the input data for the expression..

We can read it in a unit test:

```
#include <fstream>
#include "velox/vector/VectorSaver.h"

std::ifstream inputFile("/tmp/velox_vector_f7dneH", std::ifstream::binary);
auto data = restoreVector(inputFile, pool());
inputFile.close();
  
std::cout << data->toString() << std::endl;
std::cout << data->toString(0, 5) << std::endl;
```

“concat(cast((c0) as VARCHAR), ,:VARCHAR)” 
shows the expression. We can evaluate it (with minimal tweaking to replace ,:VARCHAR with ‘,’):

```
  auto result = evaluate(
      "concat(cast((c0) as VARCHAR), ',')",
      std::dynamic_pointer_cast<RowVector>(data));
```

### CODE REVIEW

File: velox/common/base/VeloxException.h [hunk 1]
```diff
@@ -55,8 +55,18 @@ struct ExceptionContext {
   /// Calls `messageFunc(arg)` and returns the result. Returns empty string if
   /// `messageFunc` is null.
   std::string message() {
-    return messageFunc ? messageFunc(arg) : "";
+    if (!suspended) {
```
Comment (pedroerp) [inline velox/common/base/VeloxException.h:58]: seems unrelated to this PR?

File: velox/vector/tests/VectorSaverTest.cpp [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+class VectorSaverTest : public testing::Test, public VectorTestBase {
+ protected:
+  void assertEqualEncodings(
+      const VectorPtr& expected,
+      const VectorPtr& actual) {
+    ASSERT_EQ(expected->encoding(), actual->encoding());
+    ASSERT_EQ(*expected->type(), *actual->type());
+    assertEqualVectors(expected, actual);
+
+    switch (expected->encoding()) {
```
Comment (pedroerp) [inline velox/vector/tests/VectorSaverTest.cpp:34]: took me a little while to understand why we also want to test this. Could use a one-liner to make it clearer for the next reader.

File: velox/vector/tests/VectorSaverTest.cpp [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+    std::ifstream inputFile(path->path, std::ifstream::binary);
+    auto copy = restoreVector(inputFile, pool());
+    inputFile.close();
+
+    // Verify encodings and data recursively.
+    assertEqualEncodings(vector, copy);
+  }
+
+  void testRoundTrip(const TypePtr& type) {
```
Comment (pedroerp) [inline velox/vector/tests/VectorSaverTest.cpp:79]: nit: `testTypeRoundTrip()` ?

File: velox/vector/tests/VectorSaverTest.cpp [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    inputFile.close();
+
+    ASSERT_EQ(*type, *copy)
+        << "Expected: " << type->toString() << ". Got: " << copy->toString();
+  }
+
+  template <typename T>
+  void testFlatIntegers() {
+    // No nulls.
+    testRoundTrip(makeFlatVector<T>({1, 2, 3, 4, 5}));
```
Comment (pedroerp) [inline velox/vector/tests/VectorSaverTest.cpp:97]: why not using the other testRoundTrip() specialization and let the fuzzer generate random input, to increase entropy?

File: velox/vector/VectorSaver.cpp [hunk 5]
```diff
@@ -0,0 +1,650 @@
...
+  if (isNull) {
+    return;
+  }
+
+  auto baseVector = vector.valueVector();
+  write<bool>(baseVector == nullptr, out);
+
+  if (baseVector) {
+    saveVector(*baseVector, out);
+    write<int32_t>(vector.wrappedIndex(0), out);
```
Comment (bikramSingh91) [inline velox/vector/VectorSaver.cpp:320]: valueVector returns the vector after removing only the constant wrapping however wrappedIndex() returns the index after removing all wrappings. Should we use index() here instead?

File: velox/vector/VectorSaver.cpp [hunk 6]
```diff
@@ -0,0 +1,650 @@
...
+  } else {
+    VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL(
+        writeConstant, vector.typeKind(), vector, out);
```
Comment (bikramSingh91) [inline velox/vector/VectorSaver.cpp:323]: nit: how about writeScalarConstant ?

**Review by bikramSingh91 — COMMENTED** [review]
Thanks for putting this together so quickly, this'll bring down the time to repro failures in expressions dramatically.
Havent finished reviewing yet, will continue next week..

File: velox/common/base/VeloxException.h (hunk 1 above)
Comment (mbasmanova) [reply velox/common/base/VeloxException.h:58]: I'm planning to use VectorSaver in the expression handing path, hence, wanted to make sure that things will be Ok if VectorSaver itself throws. Sure enough the code went into an infinite recursion of exceptions. Hence, I made that fix. I'll extract it into a separate PR though.

File: velox/vector/tests/VectorSaverTest.cpp (hunk 4 above)
Comment (mbasmanova) [reply velox/vector/tests/VectorSaverTest.cpp:97]: I'm still learning how to use the fuzzer and I find it easier to debug tests that are simply defined.

File: velox/vector/VectorSaver.cpp (hunk 5 above)
Comment (mbasmanova) [reply velox/vector/VectorSaver.cpp:320]: That's a good point. Technically speaking we should use index(). In practice though, constant vector cannot wrap neither constant nor dictionary, hence, index() will be same as wrappedIndex(). To call index(), I'll need to add a type dispatch to cast to ConstantVector<T>, since index() method is not available through BaseVector. 

```
    VELOX_CHECK_NE(
        base->BaseVector::encoding(),
        VectorEncoding::Simple::CONSTANT,
        "Constant vector cannot wrap Constant vector");
    VELOX_CHECK_NE(
        base->BaseVector::encoding(),
        VectorEncoding::Simple::DICTIONARY,
        "Constant vector cannot wrap Dictionary vector");
```

File: velox/vector/VectorSaver.cpp (hunk 6 above)
Comment (mbasmanova) [reply velox/vector/VectorSaver.cpp:323]: That's a better name. I'll change. Thank you for the suggestion.

File: velox/common/base/VeloxException.h (hunk 1 above)
Comment (mbasmanova) [reply velox/common/base/VeloxException.h]: Here is a PR: https://github.com/facebookincubator/velox/pull/2578

File: velox/vector/VectorSaver.cpp [hunk 7]
```diff
@@ -0,0 +1,652 @@
...
+
+  VELOX_FAIL("String offset is outside of the string buffers: {}", offset);
+}
+
+void writeStringViews(
+    vector_size_t size,
+    const BufferPtr& strings,
+    const std::vector<BufferPtr>& stringBuffers,
+    std::ostream& out) {
+  write<int32_t>(size * sizeof(StringView), out);
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/vector/VectorSaver.cpp:187]: That should work. Changing to `write<int32_t>(strings->size(), out);`.

File: velox/vector/tests/VectorSaverTest.cpp [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+    ASSERT_EQ(expected->encoding(), actual->encoding());
+    ASSERT_EQ(*expected->type(), *actual->type());
+    assertEqualVectors(expected, actual);
+
+    // Recursively compare inner vectors to ensure that all layers of wrappings
+    // are the same.
+    switch (expected->encoding()) {
+      case VectorEncoding::Simple::CONSTANT:
+      case VectorEncoding::Simple::DICTIONARY:
+        if (expected->valueVector()) {
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/vector/tests/VectorSaverTest.cpp:43]: Good idea. Will do.

File: velox/vector/tests/VectorSaverTest.cpp [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+  data = std::make_shared<DictionaryVector<int64_t>>(
+      pool(),
+      makeNulls(100, nullEvery(1)),
+      100,
+      BaseVector::create(BIGINT(), 0, pool()),
+      makeIndices(100, [](auto row) { return 0; }));
+  testRoundTrip(data);
+}
+
+TEST_F(VectorSaverTest, dictionaryArray) {
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/vector/tests/VectorSaverTest.cpp:367]: Good suggestion. Will add a test.

**Review by bikramSingh91 — APPROVED** [review]
